lib/teleterm/daemon/gateway_cert_reissuer.go

```diff
@@ -0,0 +1,186 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package daemon
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/sirupsen/logrus"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+
+	"github.com/gravitational/teleport/lib/client"
+	api "github.com/gravitational/teleport/lib/teleterm/api/protogen/golang/v1"
+	"github.com/gravitational/teleport/lib/teleterm/api/uri"
+	"github.com/gravitational/teleport/lib/teleterm/gateway"
+	"github.com/gravitational/teleport/lib/tlsca"
+)
+
+// GatewayCertReissuer is responsible for managing the process of reissuing a cert for a gateway
+// after the cert expires.
+type GatewayCertReissuer struct {
+	// TSHDEventsClient gets set by daemon.Service behind its mutex.
+	TSHDEventsClient TSHDEventsClient
+	// reloginMu is used when a goroutine needs to request a relogin frem the Electron app. Since the
+	// app can show only one login modal at a time, we need to submit only one request at a time.
+	reloginMu sync.Mutex
+	Log       *logrus.Entry
+}
+
+// DBCertReissuer lets us pass a mock in tests and clusters.Cluster (which makes calls to the
+// cluster) in production code.
+type DBCertReissuer interface {
+	ReissueDBCerts(context.Context, tlsca.RouteToDatabase) error
+}
+
+// TSHDEventsClient takes only those methods from api.TshdEventsServiceClient that
+// GatewayCertReissuer actually needs. It makes mocking the client in tests easier and future-proof.
+type TSHDEventsClient interface {
+	Relogin(ctx context.Context, in *api.ReloginRequest, opts ...grpc.CallOption) (*api.ReloginResponse, error)
+	SendNotification(ctx context.Context, in *api.SendNotificationRequest, opts ...grpc.CallOption) (*api.SendNotificationResponse, error)
+}
+
+// ReissueCert attempts to contact the cluster to reissue the cert used by the gateway. If that
```

Comment:
I think it can be helpful in this file's explanation comments to consistently prefix word `cert` with its qualifier (like `db cert`, `gateway cert`, `user cert`), since there's multiple certs at play and it can be confusing to keep track of 😅 
Btw, in the code we mention 3 qualifiers for a cert - db, gateway and user, but db/gateway is basically same thing?